File: llm_generated_code/Gemini/CPP/Level2/Q9.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Item {
    int weight, value;
    double ratio;
};

bool compare(Item a, Item b) {
    return a.ratio > b.ratio;
}

double fractionalKnapsack(vector<Item>& items, int capacity) {
    int n = items.size();

    // Calculate value-to-weight ratio for each item
    for (int i = 0; i < n; ++i) {
        items[i].ratio = (double)items[i].value / items[i].weight;
    }

    // Sort items by decreasing order of ratio
    sort(items.begin(), items.end(), compare);

    double totalValue = 0.0;
    int currentWeight = 0;

    // Fill the knapsack greedily
    for (int i = 0; i < n; ++i) {
        if (currentWeight + items[i].weight <= capacity) {
            currentWeight += items[i].weight;
            totalValue += items[i].value;
        } else {
            int remainingWeight = capacity - currentWeight;
            double fraction = (double)remainingWeight / items[i].weight;
            totalValue += fraction * items[i].value;
            break;
        }
    }

    return totalValue;
}

int main() {
    vector<Item> items = {{2, 10}, {3, 5}, {5, 15}, {7, 7}};
    int capacity = 8;
    double maxValue = fractionalKnapsack(items, capacity);
    cout << "Maximum value: " << maxValue << endl;

    return 0;
}

File: llm_generated_code/Gemini/Python/Level2/Q9.py
def fractional_knapsack(weights, values, capacity):
    """
    Solves the fractional knapsack problem using a greedy approach.

    Args:
        weights: A list of item weights.
        values: A list of item values.
        capacity: The maximum capacity of the knapsack.

    Returns:
        The maximum value that can be obtained.
    """

    n = len(weights)
    items = sorted(enumerate(values), key=lambda x: x[1] / weights[x[0]], reverse=True)

    total_value = 0
    current_weight = 0

    for item_index, value in items:
        weight = weights[item_index]
        if current_weight + weight <= capacity:
            total_value += value
            current_weight += weight
        else:
            remaining_capacity = capacity - current_weight
            fraction = remaining_capacity / weight
            total_value += fraction * value
            break

    return total_value

# Example usage:
weights = [2, 3, 5, 7]
values = [10, 5, 15, 7]
capacity = 8
max_value = fractional_knapsack(weights, values, capacity)
print("Maximum value:", max_value)
